Vagrantfile

```diff
@@ -436,6 +459,80 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
     override.vm.box = 'bento/ubuntu-20.04'
   end
 
+  ## Docker provider will refuse to provision these ports, even if the containers using them are turned off
+  ## Fix provided in some future version of Vagrant. https://github.com/hashicorp/vagrant/issues/13110
+  ## TODO: Review this later
+  def get_docker_used_ports
+    docker_provisioner = VagrantPlugins::DockerProvider::Driver.new
+    used_ports = Hash.new{|hash,key| hash[key] = Set.new}
+    containers = docker_provisioner.execute('docker', 'ps', '-a', '-q', '--no-trunc').to_s.split
+    containers.each do |cid|
+      container_info = JSON.parse(docker_provisioner.execute('docker', 'inspect', cid)).first
+
+      if container_info["HostConfig"]["PortBindings"]
+        port_bindings = container_info["HostConfig"]["PortBindings"]
+        next if port_bindings.empty? # Nothing defined, but not nil either
+
+        port_bindings.each do |guest_port,host_mapping|
+          host_mapping.each do |h|
+            if h["HostIp"] == ""
+              hostip = "*"
+            else
+              hostip = h["HostIp"]
+            end
+            hostport = h["HostPort"]
+            used_ports[hostport].add(hostip)
+          end
+        end
+      end
+    end
+
+    used_ports
+  end
+  
+
+  def port_is_open?(*ports)
+    used_ports = get_docker_used_ports()
+    _port = 0
+    ports.each do |port|
+      _port = port
+      next if used_ports.has_key?(_port)
+      begin
+        Socket.tcp('127.0.0.1', port, connect_timeout: 5) {}
+      rescue Errno::ECONNREFUSED, Errno::ETIMEDOUT
+        return port
+      end
+    end
+    
+    loop do
+      _port = _port + 1
+      next if used_ports.has_key?(_port)
+      begin
+        Socket.tcp('127.0.0.1', _port, connect_timeout: 5) {}
+      rescue Errno::ECONNREFUSED, Errno::ETIMEDOUT
+        return _port
+      end
+    end
+  end
+  
+  http_port = port_is_open?(80, 8080, 8888, 8000)
+  https_port = port_is_open?(443, 8443, 9999, 9000)
+  mysql_port = port_is_open?(3306)
+  mailhog_port = port_is_open?(8025)
+
+  # Docker use image.
+  config.vm.provider :docker do |d, override|
+    d.image = 'pentatonicfunk/vagrant-ubuntu-base-images:20.04'
+    d.has_ssh = true
+    d.ports =  [ "#{http_port}:80" ]
+    d.ports += [ "#{https_port}:443" ]
+    d.ports += [ "#{mysql_port}:3306" ]
+    d.ports += [ "#{mailhog_port}:8025" ]
+
+    ## Fix goodhosts aliases format for docker
+    override.goodhosts.aliases = { '127.0.0.1' => vvv_config['hosts'], '::1' => vvv_config['hosts'] }
```

Comment:
> Traefik would be useful if we wanted to have multi-container nginx, or node apps supports or something, so that you can access different things based on, for example, paths:

That is something I'd like in the future, particularly for MariaDB/ES/Mailhog. I can see an argument for there being an Nginx container too, and in the long run it also allows us to be more flexible with PHP versions and php logging on a per site basis

Reply:
also note that Altis Local Server manages this somehow with Traefik

Reply:
I can look into the code, but it's likely with one of the 3 options above.

Is the project open source for the altis solution?

Reply:
https://github.com/humanmade/altis-local-server note it's a composer plugin so it's PHP generating config files at runtime

Reply:
I don't mind us doing something similar with Ruby generating config files though, or if the eventual VVV CLI generates those files

Reply:
https://github.com/humanmade/altis-local-server/blob/f44d1ca1b803537ed5c8c940adf96ac9cfb0132b/inc/composer/class-command.php#L922

As far as I can see, they are not generating entries, at least not directly in the project.

Reply:
Even more so, they state this on [the docs](https://github.com/humanmade/altis-local-server/tree/f44d1ca1b803537ed5c8c940adf96ac9cfb0132b/docs)

> Note: Altis does not manage the host entries for subdomains or custom domains, you'll need to manage those manually, via editing /etc/hosts in Linux / macOS, or C:\Windows\System32\Drivers\etc\hosts in Windows. Altis however tries to detect if those entries do not exist, and outputs the necessary configurations to add to your hosts file.

Reply:
that's ok, I'm not seeing any crazy hacks to map it though, they're just saying go to `127.0.0.1` essentially

Reply:
In the grand scheme of things though, I don't think this is the highest priority, I'd prioritise extracting improvements that aren't strictly docker related that could be merged in advance, and having a docker file to remove the need for the pentatonic docker image so we can layer things in the future

Reply:
ok, so goodhosts stays for the foreseable future as it solves this problem

about the more prioritary things, totally onboard with doing those that can be merged in advance